libbpf-tools/profile: Add -f option to display output in folded format

ekyooo opened the pull request:
Added -f option to display output in folded format for conversion to flame graphs.
This is a test example.

```
  # ./profile -f
    test-strlen-abc;__libc_start_main;main;a;b;strlen 34
    test-strlen-abc;__libc_start_main;main;a;b;strlen 28
    test-strlen-abc;__libc_start_main;main;a;b;strlen 15
    test-strlen-abc;__libc_start_main;main;a;b;strlen 12
    test-strlen-abc;__libc_start_main;main;a;b;strlen 11
    test-strlen-abc;__libc_start_main;main;a;b;strlen 9
    test-strlen-abc;__libc_start_main;main;a;b;strlen 5
```

yonghong-song commented:
The patch includes a bunch of src/cc/libbpf changes. Please rebase on top of master and ensure the commit does not have any changes from src/cc/libbpf.

ekyooo commented:
> The patch includes a bunch of src/cc/libbpf changes. Please rebase on top of master and ensure the commit does not have any changes from src/cc/libbpf.

Oops, I'm sorry. I cleaned up changes in src/cc/libbpf and rebased to latest master branch.
Thank you.

yonghong-song commented on libbpf-tools/profile.c:
```diff
@@ -359,6 +375,50 @@ static int print_count(struct key_t *event, __u64 count, int stack_map)
 	return 0;
 }
 
+static int print_count_folded(struct key_t *event, __u64 count, int stack_map)
```
Quite some code duplication here between print_count_folded and print_count. If things need to change in the future, we need to change in two different places. I would like to add a bool parameter to print_count to indicate it whether folding is needed or not.

ekyooo replied:
Thank you for your feedback.
I updated the patch to remove duplication. Could you let me know what you think?

yonghong-song commented on libbpf-tools/profile.c:
```diff
@@ -307,53 +325,89 @@ static const char *usymname(unsigned long addr)
+	if (!folded) {
+		/* multi-line stack output */
+		ret = print_kern_stacktrace(event, stack_map, ip, fmt, false);
+		ret = print_user_stacktrace(event, stack_map, ip, fmt, ret && env.delimiter);
```
The 'ret' is not used here.

ekyooo replied:
Yes. you are right.
It was added just to look similar to the line above. But it's useless. It has now been deleted.
Thank you very much for the detailed feedback.

yonghong-song commented on libbpf-tools/profile.c:
```diff
@@ -307,53 +325,89 @@ static const char *usymname(unsigned long addr)
+		/* folded stack output */
+		printf("%s", event->name);
+		ret = print_user_stacktrace(event, stack_map, ip, fmt, false);
+		ret = print_kern_stacktrace(event, stack_map, ip, fmt, ret && env.delimiter);
```
The 'ret' is not used here.

ekyooo replied:
Thank you for this too.

yonghong-song reviewed (commented):
LGTM with two minor changes.
